coreax/coresubset.py

```diff
@@ -617,3 +615,296 @@ def _loop_body(
+        else:
+            self.stein_kernel = coreax.kernel.SteinKernel(self.kernel, score_function)
+
+        # Initialise variables that will be updated throughout the loop. These are
+        # initially local variables, with the coreset indices being assigned to self
+        # when the entire set is created
+        kernel_similarity_penalty = jnp.zeros(num_data_points)
+        try:
+            # Note that a TypeError is raised if the size input to jnp.zeros is negative
+            coreset_indices = jnp.zeros(coreset_size, dtype=jnp.int32)
+
+        except TypeError as exception:
+            if coreset_size <= 0 or isinstance(coreset_size, float):
+                raise ValueError(
+                    "coreset_size must be a positive integer"
+                ) from exception
+            raise
+
+        # Evaluate the diagonal of the Gram matrix
+        stein_kernel_diagonal = vmap(
+            self.stein_kernel.compute_elementwise, in_axes=(0, 0), out_axes=0
+        )(
+            self.original_data.pre_coreset_array,
+            self.original_data.pre_coreset_array,
+        )
+
+        regularised_log_pdf = 0.0
+
+        if self.regularise:
+            # Compute positive Laplace operator
+            def laplace_positive(x_):
+                hessian = jacfwd(score_function)(x_)
+                return jnp.clip(jnp.diag(hessian), a_min=0.0).sum()
+
+            laplace_positive_vector = vmap(laplace_positive, in_axes=(0,), out_axes=0)(
+                self.original_data.pre_coreset_array
+            )
+            # Add to kernel diagonal
+            stein_kernel_diagonal += laplace_positive_vector
+
+            # Fit a KDE to estimate log PDF: note the transpose
+            kde = gaussian_kde(
+                self.original_data.pre_coreset_array.T,
+                bw_method=float(self.kernel.length_scale),
+            )
+
+            # Use suggested regularisation parameter
+            regulariser_lambda = 1 / coreset_size
+            regularised_log_pdf = regulariser_lambda * kde.logpdf(
+                self.original_data.pre_coreset_array.T
+            )
+
+        # Greedily select coreset points
+        body = partial(
+            self._greedy_body,
+            x=self.original_data.pre_coreset_array,
+            kernel_vectorised=self.stein_kernel.compute,
+            kernel_diagonal=stein_kernel_diagonal,
+            regularised_log_pdf=regularised_log_pdf,
+            unique=self.unique,
+        )
+        try:
+            coreset_indices, kernel_similarity_penalty = lax.fori_loop(
+                lower=0,
+                upper=coreset_size,
+                body_fun=body,
+                init_val=(coreset_indices, kernel_similarity_penalty),
+            )
+        except IndexError as exception:
+            if coreset_size == 0:
+                raise ValueError("coreset_size must be non-zero") from exception
+            raise
```

Comment:
I think this try-except is also redundant due to the check on line 731

Reply:
Done.